File: src/nn.jl
using LinearAlgebra
LinearAlgebra.dot(a::Value, b::Value) = a * b

# ─── Neuron ──────────────────────────────────────────────────────────

struct Neuron
    weights::Vector{Value}
    bias::Value
    nonlin::Bool
    activation::Symbol

    function Neuron(nin; nonlin=true, activation::Symbol=:relu)
        if activation ∉ (:relu, :tanh)
            throw(ArgumentError("unsupported activation: $activation"))
        end
        return new(
        [Value(2 * rand() - 1) for _ in 1:nin],
        Value(0.0),
        nonlin,
        activation,
    )
    end
end

function (n::Neuron)(x)
    act = n.weights ⋅ x + n.bias
    if !n.nonlin
        return act
    elseif n.activation == :relu
        return relu(act)
    else
        return tanh(act)
    end
end

params(n::Neuron) = [n.weights..., n.bias]

# ─── Layer ───────────────────────────────────────────────────────────

struct Layer
    neurons::Vector{Neuron}

    Layer(nin, nout; kwargs...) = new([Neuron(nin; kwargs...) for _ in 1:nout])
end

function (l::Layer)(x)
    outs = [n(x) for n in l.neurons]
    return length(outs) == 1 ? outs[1] : outs
end

params(l::Layer) = mapreduce(params, vcat, l.neurons; init=Value[])

# ─── MLP ─────────────────────────────────────────────────────────────

struct MLP
    layers::Vector{Layer}

    function MLP(nin, nouts; activation::Symbol=:relu)
        sz = [nin; nouts]
        n = length(nouts)
        new([Layer(sz[i], sz[i+1]; nonlin=(i != n), activation=activation) for i in 1:n])
    end
end

function (m::MLP)(x)
    for layer in m.layers
        x = layer(x)
    end
    return x
end

params(m::MLP) = mapreduce(params, vcat, m.layers; init=Value[])

function zero_grad!(model)
    for param in params(model)
        param.grad = 0.0
    end
end

File: examples/karpathy/nn.py
import random

from engine import Value

ACTIVATIONS = {"relu", "tanh"}


class Module:
    def zero_grad(self):
        for p in self.parameters():
            p.grad = 0

    def parameters(self):
        return []


class Neuron(Module):
    def __init__(self, nin, nonlin=True, activation="relu"):
        if activation not in ACTIVATIONS:
            raise ValueError(f"unsupported activation: {activation}")
        self.w = [Value(random.uniform(-1, 1)) for _ in range(nin)]
        self.b = Value(0)
        self.nonlin = nonlin
        self.activation = activation

    def __call__(self, x):
        act = sum((wi * xi for wi, xi in zip(self.w, x, strict=True)), self.b)
        if not self.nonlin:
            return act
        return act.relu() if self.activation == "relu" else act.tanh()

    def parameters(self):
        return [*self.w, self.b]


class Layer(Module):
    def __init__(self, nin, nout, **kwargs):
        self.neurons = [Neuron(nin, **kwargs) for _ in range(nout)]

    def __call__(self, x):
        out = [n(x) for n in self.neurons]
        return out[0] if len(out) == 1 else out

    def parameters(self):
        return [p for n in self.neurons for p in n.parameters()]


class MLP(Module):
    def __init__(self, nin, nouts, activation="relu"):
        sz = [nin, *nouts]
        self.layers = [
            Layer(
                sz[i],
                sz[i + 1],
                nonlin=i != len(nouts) - 1,
                activation=activation,
            )
            for i in range(len(nouts))
        ]

    def __call__(self, x):
        for layer in self.layers:
            x = layer(x)
        return x

    def parameters(self):
        return [p for layer in self.layers for p in layer.parameters()]
